staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -322,101 +322,103 @@ func unescapeSingleQuote(s string) (string, error) {
-func ValidFieldPath(fieldPath string, pathOfFieldPath *field.Path, schema *schema.Structural) (validFieldPath *field.Path, err *field.Error) {
-	validFieldPath = pathOfFieldPath
-	if len(fieldPath) == 0 {
-		return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "must not be empty")
+// ValidFieldPath validates that jsonPath is a valid JSON Path containing only field and map accessors
+// that are valid for the given schema, and returns a field.Path representation of the validated jsonPath or an error.
+func ValidFieldPath(jsonPath string, schema *schema.Structural) (validFieldPath *field.Path, err error) {
+	appendToPath := func(name string) error {
+		if schema.AdditionalProperties != nil {
+			validFieldPath = validFieldPath.Key(name)
+			schema = schema.AdditionalProperties.Structural
+		} else if schema.Properties != nil {
+			validFieldPath = validFieldPath.Child(name)
+			val, ok := schema.Properties[name]
+			if !ok {
+				return fmt.Errorf("does not refer to a valid field")
+			}
+			schema = &val
+		} else {
+			return fmt.Errorf("does not refer to a valid field")
+		}
+		return nil
 	}
 
-	invalidFieldError := field.Invalid(pathOfFieldPath, fieldPath, "does not refer to an valid field")
+	validFieldPath = nil
 
-	var s scanner.Scanner
-	s.Init(strings.NewReader(fieldPath))
-	s.Filename = pathOfFieldPath.String()
-	s.Mode = scanner.ScanInts | scanner.ScanIdents | scanner.ScanChars | scanner.ScanStrings
-	s.Error = func(s *scanner.Scanner, msg string) {
-		field.Invalid(pathOfFieldPath, fieldPath, fmt.Sprintf("failed to parse JSON Path: %s", msg))
-	}
+	scanner := bufio.NewScanner(strings.NewReader(jsonPath))
 
-	found := false
-	for true {
-		tok := s.Scan()
-		if tok == scanner.EOF {
-			found = true
-			break
-		}
-		switch schema.Type {
-		case "object":
-			isMapSyntax := false
-			if s.TokenText() == "[" {
-				isMapSyntax = true
-			} else if s.TokenText() != "." {
-				return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "expected [ or . but got: "+s.TokenText())
-			}
-
-			tok = s.Scan()
-			if tok == scanner.EOF {
-				return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "unexpected end of JSON path")
-			}
-			fieldName := s.TokenText()
-			if isMapSyntax {
-				if tok == scanner.Char {
-					newS := fieldName[1 : len(fieldName)-1]
-					newS, err := unescapeSingleQuote(newS)
-					if err != nil {
-						return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, fmt.Sprintf("failed to unescape: %v", err))
+	// configure the scanner to split the string into tokens.
+	// The three delimiters ('.', '[', ']') will be returned as single char tokens.
+	// All other text between delimiters is returned as string tokens.
+	scanner.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		if len(data) > 0 {
+			for i := 0; i < len(data); i++ {
+				// If in a single quoted string, look for the end of string
+				// ignoring delimiters.
+				if data[0] == '\'' {
```

Comment:
should the single-quoting/escaping handling only trigger inside []? I can't tell whether this will do weird things with `.'foo'bar`

Reply:
We could add isQuotedString and isIdentifier utility functions to enforce that we report better errors if the wrong token is used in either place.  Right now the parser does check for quoted strings for [] but doesn't check for raw identifiers with . So the error message will be produced when the parser can't find a matching schema property.

Reply:
I have added test cases for this one. My understanding is that `.'foo'bar` should be treated as invalid fieldPath while `"['\\'foo\\'bar']"` should be valid